### PR TITLE
Add GraphQL support to API Docs 

### PR DESCRIPTION
**Pull request by andrewthauer**
## Hey, I just made a Pull Request!

This is a quick & rough experiment for adding support for `graphql` API types.

- Using `graphiql` for now. Most solutions for generating docs from graphql SDL schemas are tailored for generating static sites (e.g. https://github.com/2fd/graphdoc) and not dynamically. `GraphiQL` has a docs explorer but it's not ideal for documentation only. If anyone knows of something else that would be great.
- Styling is really rough (borrowed from `plugins/graphiql`)

image

image

image

image

#### Future

- Loading the API spec definition from an external location would be ideal. Possibly this could be done with locations (e.g. #2001)
- Would be interesting to extend the API catalog model to support annotations to indicate the URL to a live instance (possibly including params for any required headers, etc.).

#### :heavy_check_mark: Checklist

- [ ] All tests are passing `yarn test`
- [ ] Screenshots attached (for UI changes)
- [ ] Relevant documentation updated
- [ ] Prettier run on changed files
- [ ] Tests added for new functionality
- [ ] Regression tests added for bug fixes

### CODE REVIEW

File: plugins/api-docs/package.json [hunk 1]
```diff
@@ -45,6 +46,7 @@
     "@testing-library/user-event": "^12.0.7",
     "@types/jest": "^26.0.7",
     "@types/node": "^12.0.0",
+    "@types/react": "^16.9",
```
Comment (freben) [inline plugins/api-docs/package.json:49]: see the build notification - this actually should go up in the normal dependencies :)

**Review by freben — APPROVED** [review]
We can go with this for now, sure! :)

**Comment by freben** [comment]
Eventually we may want to use something like prismjs, which seems to have graphql support https://prismjs.com/#supported-languages

**Comment by andrewthauer** [comment]
> Eventually we may want to use something like prismjs, which seems to have graphql support https://prismjs.com/#supported-languages

@freben - That's actually a great idea for the `PlainApiDefinitionWidge`, I created #2170 🎉

**Comment by freben** [comment]
Hey again! Now with #2170, where do we go with this PR? :) Would you like to switch to prism or get this in as-is?

**Comment by andrewthauer** [comment]
Hi @freben - Sorry, been focused on other things.

This started as a quick experiment to see what might be possible. I did implement #2170 which potentially added syntax highlighting using the `CodeSnippet` component. I thought it was actually using `prism` under the covers but it is using `hljs` which is the default for `react-syntax-highlighter` and doesn't include `graphql` syntax (even thought it looks decent :). The `CodeSnippet` component could in theory be updated to use `prism` as it is supported by `react-syntax-highlighter`. However, that likely would break styling so opting for a new component might be better.

Anyways, that aside, I'm open to merging then once I fix the styling, but I'm not hugely fond of this particular setup. I would have preferred if you could extract or make the docs more prominent with `graphiql` but they API seems limited. There is a complete revamp happening for v2 which could make this more feasible. I haven't been able to find any `graphql` react based dynamic docs specific viewers unfortunately.

Another option could be to have the `API` entity support something like an external URL reference instead of an inline `definition` and then have an embedded iframe or use shadow dom to render the external API. Internally we use [graphdoc](https://github.com/2fd/graphdoc) right now which generates the side dynamically. This could be a more general purpose option as well.

cc @Fox32, @dhenneke

**Comment by andrewthauer** [comment]
@freben - I updated the styling so the graphiql takes up 100% height. However, to get this to work and look relatively decent I needed to modify the padding for the containers. It might actually be better if the individual api definition's themselves provided the padding to avoid this contention. Screenshots updated in description of this PR.

**Comment by freben** [comment]
Hey! I felt so bad for merging PRs in the wrong order - a younger one got in first and now it conflicted with yours - I took the liberty of rebasing and fixing the conflicts and pushing a branch to our repo with the same name as yours https://github.com/spotify/backstage/compare/api-docs-graphql

Feel free to hard-reset to that one or so to save time. Sorry for the noise :)

**Comment by andrewthauer** [comment]
@freben - No problem at all. I actually need to confirm this also doesn't break support for displaying multiple implemented apis for a single component (e.g. `openapi` + `graphql`, etc.).

**Comment by dhenneke** [comment]
> Another option could be to have the API entity support something like an external URL reference instead of an inline definition and then have an embedded iframe or use shadow dom to render the external static generated API docs. Internally we use graphdoc right now which generates the side dynamically. This could be a more general purpose option as well.

I'm not sure if we want that. It can be useful to have the raw API description/schema for other means (linting, indexing/searching, etc).

But it might be a good idea to make [`ApiDefinitionWidget`](https://github.com/spotify/backstage/blob/2b114fde44793a93e09311b11233e6be2cc1b2ac/plugins/api-docs/src/components/ApiDefinitionWidget/ApiDefinitionWidget.tsx#L27) configurable in the future, or even move it from the plugin to the app. This way each instance can compose its own display for the individual APIs. Maybe one wants to use GraphiQL while others just want to show a link/iframe to an externally hosted documentation. Or maybe one want to display special annotations from the Component->API relationship (after #1964 has landed) like an API endpoint, authorization/access scope, …. The api-docs plugin then provides common default widgets that can be exchanged with custom ones if needed.

**Comment by andrewthauer** [comment]
> > Another option could be to have the API entity support something like an external URL reference instead of an inline definition and then have an embedded iframe or use shadow dom to render the external static generated API docs. Internally we use graphdoc right now which generates the side dynamically. This could be a more general purpose option as well.
> 
> I'm not sure if we want that. It can be useful to have the raw API description/schema for other means (linting, indexing/searching, etc).
> 
> But it might be a good idea to make [`ApiDefinitionWidget`](https://github.com/spotify/backstage/blob/2b114fde44793a93e09311b11233e6be2cc1b2ac/plugins/api-docs/src/components/ApiDefinitionWidget/ApiDefinitionWidget.tsx#L27) configurable in the future, or even move it from the plugin to the app. This way each instance can compose its own display for the individual APIs. Maybe one wants to use GraphiQL while others just want to show a link/iframe to an externally hosted documentation. Or maybe one want to display special annotations from the Component->API relationship (after #1964 has landed) like an API endpoint, authorization/access scope, …. The api-docs plugin then provides common default widgets that can be exchanged with custom ones if needed.

@dhenneke - This makes sense to me. I also don't like having to add hints into the API entity spec to drive the UI specifically. Having the ability for people to swap out the viewer of choice for a particular API definition makes most sense (especially if the `api-docs` plugin provides the components to compose them easily).

File: plugins/api-docs/src/components/GraphQlDefinitionWidget/GraphQlDefinitionWidget.tsx [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+export const GraphQlDefinitionWidget = ({ definition }: Props) => {
+  const classes = useStyles();
+  const schema = buildSchema(definition);
+
+  return (
+    <Suspense fallback={<Progress />}>
+      <div className={classes.root}>
+        <div className={classes.graphiQlWrapper}>
+          <GraphiQL
```
Comment (Rugvip) [inline plugins/api-docs/src/components/GraphQlDefinitionWidget/GraphQlDefinitionWidget.tsx:56]: Wonder if we can get the `graphiql` plugin in here somehow :grin:

This makes sense for now anyway, 👍  for dynamic import

**Review by Rugvip — APPROVED** [review]

**Pull request merged**